### PR TITLE
restore missing translation keys

### PR DESCRIPTION
**Pull request by tomlev**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.4.x
| Description?  | restore missing translation keys
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | 

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9183)

### CODE REVIEW

File: src/PrestaShopBundle/Translation/KeysReference/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+This directory aims to get all translation keys that need to be referenced, but not cearly visible in code.
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Translation/KeysReference/README.md]: cearly? what does it means?

File: src/PrestaShopBundle/Translation/KeysReference/XLang/AttributeGroupLang.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Translation\KeysReference\Xlang;
+
+use PrestaShopBundle\Translation\TranslatorComponent as Translator;
+
+class AttributeGroupLang
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Translation/KeysReference/XLang/AttributeGroupLang.php]: Would you mind to add PHPDoc for classes?

**Review by mickaelandrieu — CHANGES_REQUESTED** [review]

**Comment by mickaelandrieu** [comment]
I have multiples concerns with this pull request:

* why PHP files and not xliff files directly? This looks so dirty :/
* I don't understand how translator is found in every class, I have the feeling that theses classes shouldnt be instanciated at all, am I right?
* finally, I don't think these files should be placed in the bundle. Maybe in Core or better: in a new package called "missing-translations" or why not directly on TranslationsToolsBundle?

File: src/PrestaShopBundle/Translation/KeysReference/XLang/AttributeGroupLang.php (hunk 2 above)
Comment (tomlev) [reply src/PrestaShopBundle/Translation/KeysReference/XLang/AttributeGroupLang.php]: there was none before being deleted, I don't really know each class meaning, here we just need translation to be kept back

**Comment by mickaelandrieu** [comment]
Other questions: 

* do these files/translation keys are going to change over time or it is something we need to keep and never change again?
* is it possible to get these translations in a more common format? I know we can parse Xliff files really easy, with a minor update of translation tools.
* why not replace the translations where they were in the past? https://github.com/PrestaShop/PrestaShop/pull/8696

File: src/PrestaShopBundle/Translation/KeysReference/README.md (hunk 1 above)
Comment (PierreRambaud) [reply src/PrestaShopBundle/Translation/KeysReference/README.md]: That means translation tool need to fetch `trans()` method to retrieve and store translations (what a mess :smile: ): https://github.com/PrestaShop/TranslationToolsBundle/blob/master/Translation/Extractor/Visitor/TranslationNodeVisitor.php

**Comment by tomlev** [comment]
1/ translation extraction should not delete some old keys : they can be used by modules.
this change will be addressed later
2/ these translation keys cannot be replaced in their original location, because it was not clean to mix business behavior and translation-specific storage (that's the reason I deleted these keyx without doubt ^^)
3/ about the format, the goal of this PR was simply to fix the missing keys, but we can add it the change of the translation tool if needed

**Comment by mickaelandrieu** [comment]
Thanks for the answers, I'm in favor of puting these files in TranslationTools app then, in a specific folder and then to add this folder to the paths the command used to extract translations. In a second time, to define how we can introduce again these keys in the core: wdyt?

**Comment by tomlev** [comment]
I'm not sure putting these classes out of the core is a good idea. This will make things hard to maintain...
I prefer attaching them in a specific folder in the core

File: src/PrestaShopBundle/Translation/KeysReference/README.md [hunk 3]
```diff
@@ -0,0 +1,2 @@
+This directory aims to get all translation keys that need to be referenced, but not clearly visible in code.
+The translation parser will be grateful to find such readable keys, and will not delete them from the translation system
```
Comment (eternoendless) [inline src/PrestaShopBundle/Translation/KeysReference/README.md]: > The translation parser will be grateful to find such readable keys

😂

File: src/PrestaShopBundle/Translation/KeysReference/XLang/FeatureLang.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    protected function init()
+    {
+         $this->translator->trans('Height', array(), 'Shop.Demo.Catalog', $this->locale);
+         $this->translator->trans('Width', array(), 'Shop.Demo.Catalog', $this->locale);
+         $this->translator->trans('Depth', array(), 'Shop.Demo.Catalog', $this->locale);
+         $this->translator->trans('Weight', array(), 'Shop.Demo.Catalog', $this->locale);
+         $this->translator->trans('Compositions', array(), 'Shop.Demo.Catalog', $this->locale);
+         $this->translator->trans('Styles', array(), 'Shop.Demo.Catalog', $this->locale);
+         $this->translator->trans('Properties', array(), 'Shop.Demo.Catalog', $this->locale);
```
Comment (eternoendless) [inline src/PrestaShopBundle/Translation/KeysReference/XLang/FeatureLang.php]: I don't think this class is needed anymore, demo data is supposedly translated on the fixture files, and there's no more untranslation/retranslation, isn't it?

File: src/PrestaShopBundle/Translation/KeysReference/XLang/FeatureValueLang.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+         $this->translator->trans('0.41 in (clip included)', array(), 'Shop.Demo.Catalog', $this->locale);
+
+         $this->translator->trans('4.33 in', array(), 'Shop.Demo.Catalog', $this->locale);
+
+         $this->translator->trans('2.76 in', array(), 'Shop.Demo.Catalog', $this->locale);
+
+         $this->translator->trans('120g', array(), 'Shop.Demo.Catalog', $this->locale);
+
+         $this->translator->trans('0.31 in', array(), 'Shop.Demo.Catalog', $this->locale);
```
Comment (eternoendless) [inline src/PrestaShopBundle/Translation/KeysReference/XLang/FeatureValueLang.php]: Same goes for this class

File: src/PrestaShopBundle/Translation/KeysReference/XLang/GenderLang.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    /** @var Translator  */
+    protected $translator;
+
+    /** @var string */
+    protected $locale;
+
+    protected function init()
+    {
+         $this->translator->trans('Mr.', array(), 'Admin.Shopparameters.Feature', $this->locale);
+         $this->translator->trans('Mrs.', array(), 'Admin.Shopparameters.Feature', $this->locale);
```
Comment (eternoendless) [inline src/PrestaShopBundle/Translation/KeysReference/XLang/GenderLang.php]: And this class

File: src/PrestaShopBundle/Translation/KeysReference/XLang/FeatureLang.php (hunk 4 above)
Comment (tomlev) [reply src/PrestaShopBundle/Translation/KeysReference/XLang/FeatureLang.php]: I don't know how demo data is handled, and there is still untranslation/retranslation ;)

**Comment by eternoendless** [comment]
I'm not sure how TranslationTool discovers the wordings, i.e I don't know if it is enough for them just to be there or what.

@PierreRambaud, @Quetzacoalt91 what do you guys think?

In any case, I'd prefer for them to be in a configuration folder somewhere, we do have a lot to choose from.

File: src/PrestaShopBundle/Translation/KeysReference/XLang/CmsCategoryLang.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+{
+    /** @var Translator  */
+    protected $translator;
+
+    /** @var string */
+    protected $locale;
+
+    protected function init()
+    {
+         $this->translator->trans('Home', array(), 'Admin.Catalog.Feature', $this->locale);
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Translation/KeysReference/XLang/CmsCategoryLang.php]: Useless class duplicate from CategoryLang

**Review by PierreRambaud — DISMISSED** [review]
Useless class :)

**Comment by eternoendless** [comment]
According to some tests run by @PierreRambaud, we can put the files pretty much anywhere as long as it's within one of the following directories:

- classes
- controllers
- admin-dev
- modules
- src
- override
- install-dev
- themes
- pdf

Additionally, the file doesn't even need to be a class, this exact code snippet would work as well:

```php
<?php

trans('test', 'Test.test');
```

**Comment by PierreRambaud** [comment]
Directories can be configurable if we update the `translationtool` project :)

**Comment by mickaelandrieu** [comment]
@PierreRambaud good to know :)

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Thanks @tomlev !

**Pull request merged**